attest/core/src/ias/verifier.rs

```diff
@@ -0,0 +1,1138 @@
+pub enum VerifierError {
+    /// The IAS report does not contain a certificate chain.
+    NoChain,
+    /// The signature is invalid, or was produced by a public key we do not
+    /// trust.
+    BadSignature,
+    /// There was an error parsing the JSON contents
+    Parse(VerifyError),
+    /// The report was properly constructed did not meet security requirements.
+    Verification(VerificationReportData),
+}
+
+impl From<VerifyError> for VerifierError {
+    fn from(src: VerifyError) -> VerifierError {
+        VerifierError::Parse(src)
+    }
+}
+
+const MAX_CHAIN_DEPTH: usize = 5;
+
+impl IasReportVerifier {
+    /// Verify the given IAS report using this verifier object.
+    pub fn verify(
+        &self,
+        report: &VerificationReport,
+    ) -> Result<VerificationReportData, VerifierError> {
+        // Here's the background information for this code:
+        //
+        //  1. An X509 certificate can be signed by only one issuer.
+        //  2. mbedtls' certificates-list API demands certs in the RFC5246
+        //     order (endpoint cert first, every other cert signed the
+        //     cert preceeding it in the list).
+        //  3. I don't recall Intel's specification mentioning certificate
+        //     ordering at all (meaning they can change it w/o warning).
+        //  4. mbedtls' certificates-list API isn't actually exposed to us,
+        //     anyways.
+        //
+        // As a result, we need to find the cert which signed the data (this
+        // doubles as the signature check), then find a way back up the
+        // derived chain until we either hit a max-height limit (and fail),
+        // or top out at something that was itself signed by a trust_anchor.
+        //
+        // If we have the root CA that's in our trust_anchor list in the
+        // provided chain, then it will pass the "signed by trust_anchor"
+        // check, because all root CAs are self-signed by definition.
+        //
+        // If Intel doesn't provide the root CA in the chain, then the last
+        // entry in the derived chain will still contain the intermediate CA,
+        // which will (obviously) be signed by the root CA. Combined, these
+        // two situations mean checking that the last cert in the list was
+        // signed by a trust anchor will result in success.
+        //
+        // The third possible scenario, which is that one of the certs in the
+        // middle of the chain is in our trust_anchors list. In this case, our
+        // explicit trust of a cert makes any other issuer relationships
+        // irrelevant, including relationships with blacklisted issuers.
+        //
+        // This scenario is less likely, but would occur when someone is
+        // trying to deprecate an existing authority in favor of a new one. In
+        // this case, they sign the old CA with the new CA, so things which
+        // trust the new CA also trust the old CA "for free". When everyone
```

Comment:
is it not the other way around? why dont you sign the new one with the old one?

in the case that the old CA is believed to be compromised then i think you dont want to sign the old CA with the new CA? because then there is never any trust relationship such that trusting new -> trusting old
maybe it doesn't matter though

Reply:
This is a hypothetical scenario: Intel hasn't published anything (AFAIK) about the circumstances under which they'll deprecate a trust anchor (i.e. more than just compromise), or how they'll handle the transition from old to new.